include/stats.h

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright Runtime.io 2018. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Statistics.
+ *
+ * Statistics are per-module event counters for troubleshooting, maintenance,
+ * and usage monitoring.  Statistics are organized into named "groups", with
+ * each group consisting of a set of "entries".  An entry corresponds to an
+ * individual counter.  Each entry can optionally be named if the STATS_NAMES
+ * setting is enabled.  Statistics can be retrieved with the mcumgr management
+ * subsystem.
+ *
+ * There are two, largely duplicated, statistics sections here, in order to
+ * provide the optional ability to name statistics.
+ *
+ * STATS_SECT_START/END actually declare the statistics structure definition,
+ * STATS_SECT_DECL() creates the structure declaration so you can declare
+ * these statistics as a global structure, and STATS_NAME_START/END are how
+ * you name the statistics themselves.
+ *
+ * Statistics entries can be declared as any of several integer types.
+ * However, all statistics in a given structure must be of the same size, and
+ * they are all unsigned.
+ *
+ * - STATS_SECT_ENTRY(): default statistic entry, 32-bits.
+ *
+ * - STATS_SECT_ENTRY16(): 16-bits.  Smaller statistics if you need to fit into
+ *   specific RAM or code size numbers.
+ *
+ * - STATS_SECT_ENTRY32(): 32-bits.
+ *
+ * - STATS_SECT_ENTRY64(): 64-bits.  Useful for storing chunks of data.
+ *
+ * Following the static entry declaration is the statistic names declaration.
+ * This is compiled out when the CONFIGURE_STATS_NAME setting is undefined.
+ *
+ * When CONFIG_STATS_NAMES is defined, the statistics names are stored and
+ * returned to the management APIs.  When the setting is undefined, temporary
+ * names are generated as needed with the following format:
+ *
+ *     s<stat-idx>
+ *
+ * E.g., "s0", "s1", etc.
+ */
+
+#ifndef H_STATS_
+#define H_STATS_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct stats_name_map {
+	u16_t snm_off;
+	const char *snm_name;
+} __packed;
+
+struct stats_hdr {
+	const char *s_name;
+	u8_t s_size;
+	u8_t s_cnt;
+	u16_t s_pad1;
+#ifdef CONFIG_STATS_NAMES
+	const struct stats_name_map *s_map;
+	int s_map_cnt;
+#endif
+	struct stats_hdr *s_next;
+};
+
+/**
+ * @brief Declares a stat group struct.
+ *
+ * @param group__               The name to assign to the structure tag.
+ */
+#define STATS_SECT_DECL(group__) \
+	struct stats_ ## group__
+
+/**
+ * @brief Ends a stats group struct definition.
+ */
+#define STATS_SECT_END };
```

Comment:
maybe this can be changed to `#define STATS_SECT_END() }`
so it would be used as  STATS_SECT_END();

this will mitigate checkpath issue.

Reply:
obviously then need to align other STATS_xx_END(*) to this